model/patch/github.go

```diff
@@ -0,0 +1,131 @@
+package patch
+
+import (
+	"strings"
+
+	"github.com/evergreen-ci/evergreen/db"
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/pkg/errors"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const (
+	// IntentCollection is the database collection that stores patch intents.
+	IntentCollection = "patch_intent"
+
+	// GithubIntentType represents patch intents created for GitHub.
+	GithubIntentType = "github"
+)
+
+// Intent represents an intent to create a patch build and is processed by an amboy queue.
+type Intent interface {
+	// Insert inserts a patch intent in the database.
+	Insert() error
+
+	// SetProcessed should be called by an amboy queue after creating a patch from an intent.
+	SetProcessed() error
+
+	// IsProcessed returns whether a patch exists for this intent.
+	IsProcessed() bool
+
+	// GetType returns the patch intent, e.g., GithubType.
+	GetType() string
+}
+
+// githubIntent represents an intent to create a patch build as a result of a
+// PullRequestEvent webhook. These intents are processed asynchronously by an
+// amboy queue.
+type githubIntent struct {
+	// ID is created by the driver and has no special meaning to the application.
+	ID bson.ObjectId `bson:"_id"`
+
+	// PRNumber is the PR number for the project in GitHub.
+	PRNumber int `bson:"pr_number"`
+
+	// HeadSHA is the base SHA of the patch.
+	HeadSHA string `bson:"head_sha"`
+
+	// URL is the URL of the patch in GitHub.
+	URL string `bson:"url"`
+
+	// Processed indicates whether a patch intent has been processed by the amboy queue.
+	Processed bool `bson:"processed"`
+
+	// IntentType indicates the type of the patch intent, i.e., GithubIntentType
+	IntentType string `bson:"intent_type"`
+}
+
+// BSON fields for the patches
+var (
+	idKey         = bsonutil.MustHaveTag(githubIntent{}, "ID")
+	prNumberKey   = bsonutil.MustHaveTag(githubIntent{}, "PRNumber")
+	headSHAKey    = bsonutil.MustHaveTag(githubIntent{}, "HeadSHA")
+	urlKey        = bsonutil.MustHaveTag(githubIntent{}, "URL")
+	processedKey  = bsonutil.MustHaveTag(githubIntent{}, "Processed")
+	intentTypeKey = bsonutil.MustHaveTag(githubIntent{}, "IntentType")
+)
+
+// NewGithubIntent return a new github patch intent.
+func NewGithubIntent(pr int, sha string, url string) (*githubIntent, error) {
```

Comment:
public functions that return private types are awkward. 

we're also not using the interface anywhere yet, what's the plan there?

Reply:
> public functions that return private types are awkward.

Made it public. I had it public in the first place :-(.

> we're also not using the interface anywhere yet, what's the plan there?

The plan is to define a function that returns an amboy job from the patch_intent. Thinking about this more, I think we will remove `IntentType` and have a different function, but I think this is fine for now.